### PR TITLE
Assignment - Nitesh Agarwal - Movie Search

### PR DESCRIPTION
**Pull request by agarwalnitesh42**
Changes Done
---
1. Code Restructuring on models, constants etc..
2. A Custom Pipe Created for filtering purpose, so existing filter  related code have been removed.
3. Optimized and put some readibility in data.service.ts
4. Instead of using subscription/unsubscription. Used async which handles unsubscription by itself. so related codes on ngOnDestroy won't be required now , hence removed and made other necessary changes in html files.
5. Bugs Fixes for run time changes
6. Layout fixes
7. Mock data separated in a different folder to avoid rewriting in every spec file
8. Some Adhoc Changes and Improvements.
9. WIP - Still Need to make few more changes/improvements in Spec Files. 

@ThomasBurleson  @jasonruesch @tscott42 Kindly Review the PR. 

Note: Since Last 3 years , I have been working on React/React Native , So still need to brush up my Angular Skills. However during this assignment. I went through basics theories again to start off the assignment.

### CODE REVIEW

File: FE/angular/src/app/services/data.service.ts [hunk 1]
```diff
@@ -83,27 +33,41 @@ export class DataService {
...
+    return this.http.get<SearchResults>(`${APP_CONSTANTS.serviceUrl}s=Batman&type=movie`).pipe(
+      mergeMap(({ Search }) => {
+        const imdbIDs = Search.map(({ imdbID }) => imdbID);
+        return this.fetchMoviesDetails(imdbIDs).pipe(
+          map((movies) => {
+            const decadesSet = new Set<number>();
+
+            // Calculate decades and eliminate duplicates
+            movies.forEach((movie: Movie) => {
+              const decade = Math.ceil(parseInt(movie.Year.toString(), 10) / 10) * 10 - 10;
```
Comment (tscott42) [inline FE/angular/src/app/services/data.service.ts:45]: can this be simplified, do we need to use .toString() here?

File: FE/angular/src/app/services/data.service.ts [hunk 2]
```diff
@@ -83,27 +33,41 @@ export class DataService {
...
+              decadesSet.add(decade);
+            });
 
-            return this.getMovie(imdbID);
+            // Sort the decades
+            const decades = Array.from(decadesSet).sort((a, b) => a - b);
+
+            // Sort movies by year
+            const sortedMovies = movies.sort((a, b) => parseInt(a.Year.toString(), 10) - parseInt(b.Year.toString(), 10));
```
Comment (tscott42) [inline FE/angular/src/app/services/data.service.ts:53]: can this be simplified, do we need to use .toString() here?

File: FE/angular/src/app/components/movies/movies.component.spec.ts [hunk 3]
```diff
@@ -1,113 +1,34 @@
...
-    Plot: 'Mock movie plot summary.',
-    Poster:
-      'https://m.media-amazon.com/images/M/MV5BOTY4YjI2N2MtYmFlMC00ZjcyLTg3YjEtMDQyM2ZjYzQ5YWFkXkEyXkFqcGdeQXVyMTQxNzMzNDI@._V1_SX300.jpg',
-    imdbID: 'tt123',
-    Type: 'movie'
-  }
-];
+describe('MoviesComponent', () => {
+  let component: MoviesComponent;
+  let fixture: ComponentFixture<MoviesComponent>;
```
Comment (tscott42) [inline FE/angular/src/app/components/movies/movies.component.spec.ts:11]: Curious why you chose to use TestBed instead of Spectator

**Review by tscott42 — COMMENTED** [review]
Any thoughts on how to improve the UI when the movie poster image is not available or when the movie title is longer than one line?

File: FE/angular/src/app/components/movies/movies.component.spec.ts [hunk 4]
```diff
@@ -1,113 +1,34 @@
...
+    component = fixture.componentInstance;
   });
 
-  test('should create the component', () => {
-    component.ngOnInit();
+  it('should create the component', () => {
     expect(component).toBeTruthy();
   });
 
-  describe('ngOnInit', () => {
```
Comment (ThomasBurleson) [inline FE/angular/src/app/components/movies/movies.component.spec.ts:72]: Why did you delete all these tests ?

File: FE/angular/src/app/components/movie/movie.component.ts [hunk 5]
```diff
@@ -1,25 +1,23 @@
...
   public ngOnInit() {
-    this.activatedRoute.params.pipe(tap(({ id }) => (this.movieId = id)));
-    this.movieSubscription = this.dataService.getMovie(this.movieId).pipe(tap((data) => (this.movie = data)));
-  }
-
-  public ngOnDestroy(): void {
-    this.movieSubscription.unsubscribe();
+    this.movie$ = this.activatedRoute.params.pipe(
+      tap(({ id }) => (this.movieId = id)),
+      switchMap(({ id }) => this.dataService.getMovie(id))
```
Comment (ThomasBurleson) [inline FE/angular/src/app/components/movie/movie.component.ts:20]: Does this not have Lint issues?

File: FE/angular/src/app/components/movies/movies.component.ts [hunk 6]
```diff
@@ -1,41 +1,30 @@
-import { Component, OnDestroy, OnInit } from '@angular/core';
-import { tap } from 'rxjs';
-import { DataService, MovieComplete } from '../../services/data.service';
+import { Component, OnInit } from '@angular/core';
+import { Observable, map } from 'rxjs';
+import { DataService } from '../../services/data.service';
+import { MovieComplete } from '../../models';
+import { FilterByDecadePipe } from '../../pipes/filterByDecade.pipe';
```
Comment (ThomasBurleson) [inline FE/angular/src/app/components/movies/movies.component.ts:5]: Why is this imported ?

File: FE/angular/src/app/components/movies/movies.component.html [hunk 7]
```diff
@@ -2,11 +2,11 @@
...
   [decades]="decades"
   [currDecade]="currDecade"
-  (updateDecade)="displayMovies($event)"
+  (updateDecade)="updateCurrDecade($event)"
 ></app-decades>
 
-<ul class="movies" *ngIf="filteredMovies.length">
-  <li *ngFor="let movie of filteredMovies" class="grid grid-bleed">
+<ul class="movies" *ngIf="movies$ | async as movies; else loadingMovies">
+  <li *ngFor="let movie of movies | filterByDecade: currDecade" class="grid grid-bleed">
```
Comment (ThomasBurleson) [inline FE/angular/src/app/components/movies/movies.component.html:9]: This is creative using a pipe to filter the movies. :-)
Be prepared to talk about the pros/cons of this. 
Is this pipe reusable?

File: FE/angular/src/app/components/movies/movies.component.ts [hunk 8]
```diff
@@ -1,41 +1,30 @@
...
   public currDecade: number | undefined;
   public decades: number[] = [];
-  public filteredMovies: MovieComplete[] = [];
-  public movies: MovieComplete[] = [];
-  private moviesSubscription: any;
+  public movies$: Observable<MovieComplete[]> | null = null;
 
   constructor(private dataService: DataService) {}
 
   public ngOnInit(): void {
```
Comment (ThomasBurleson) [inline FE/angular/src/app/components/movies/movies.component.ts:18]: Why are you doing this in `ngOnInit()`

File: FE/angular/src/app/constants/index.ts [hunk 9]
```diff
@@ -0,0 +1,7 @@
+const APP_CONSTANTS = {
```
Comment (ThomasBurleson) [inline FE/angular/src/app/constants/index.ts:1]: Nice way to centralize constants?
Why `APP_CONSTANTS`? Would there be a better name?
What if we wanted these to be Dependency Injected? (how would you do this)?

File: FE/angular/src/app/models/index.ts [hunk 10]
```diff
@@ -0,0 +1,33 @@
+export interface SearchResults {
```
Comment (ThomasBurleson) [inline FE/angular/src/app/models/index.ts:1]: What are `index.ts` files used for ? (what is another name for this file)?
Why is defining these interfaces here a bad idea?

File: FE/angular/src/app/pipes/filterByDecade.pipe.ts [hunk 11]
```diff
@@ -0,0 +1,16 @@
+import { Pipe, PipeTransform } from '@angular/core';
+import { MovieComplete } from '../models';
+
+@Pipe({
+  name: 'filterByDecade',
+})
+export class FilterByDecadePipe implements PipeTransform {
+  transform(movies: MovieComplete[], decade?: number): MovieComplete[] {
+    if (!decade) {
```
Comment (ThomasBurleson) [inline FE/angular/src/app/pipes/filterByDecade.pipe.ts:9]: What if `decade` is a negative number?

File: FE/angular/src/app/services/data.service.ts [hunk 12]
```diff
@@ -83,27 +33,41 @@ export class DataService {
...
+        console.error('HTTP request error:', error);
+        return throwError(error); // Rethrow the error
+      }),
+      shareReplay(1) // Cache the result to avoid making the same HTTP request multiple times
+    );
+  }
 
-        return this.storedMovies;
-      })
+  private fetchMoviesDetails(imdbIDs: string[]): Observable<MovieComplete[]> {
```
Comment (ThomasBurleson) [inline FE/angular/src/app/services/data.service.ts:68]: Nice.

File: FE/angular/src/app/services/data.service.ts [hunk 13]
```diff
@@ -1,80 +1,30 @@
...
-        Rated,
-        Released,
-        Runtime,
-        Title,
-        Type,
-        Writer,
-        Year: parseInt(Year as string)
-      }))
+  getMovie(id: string): Observable<MovieComplete> {
+    return this.http.get<MovieDetails>(`${APP_CONSTANTS.serviceUrl}i=${id}`).pipe(
```
Comment (ThomasBurleson) [inline FE/angular/src/app/services/data.service.ts:18]: Nice.

File: FE/angular/src/app/services/data.service.ts [hunk 14]
```diff
@@ -83,27 +33,41 @@ export class DataService {
...
-        Search = Search.sort(({ Year: year1 }: MovieComplete, { Year: year2 }: MovieComplete) => year1 - year2);
-        this.decades.sort((a, b) => a - b);
-        this.storedMovies = { Search, Decades: this.decades };
+        );
+      }),
+      catchError((error) => {
+        console.error('HTTP request error:', error);
+        return throwError(error); // Rethrow the error
+      }),
+      shareReplay(1) // Cache the result to avoid making the same HTTP request multiple times
```
Comment (ThomasBurleson) [inline FE/angular/src/app/services/data.service.ts:64]: Is this needed?
